# Práctica 9

In [198]:
import pandas as pd
import numpy as np

In [199]:
satellites_df = pd.read_excel('satellites.xls')
print(satellites_df.shape)
satellites_df.head()

(3375, 67)


Name of Satellite, Alternate Names  \
0  1HOPSAT-TD (1st-generation High Optical Perfor...   
1                                             3Cat-1   
2                                           Aalto-1    
3                                           AAUSat-4   
4                           ABS-2 (Koreasat-8, ST-3)   

  Current Official Name of Satellite Country/Org of UN Registry  \
0                            1HOPSAT                 NR (12/20)   
1                             3Cat-1                         NR   
2                            Aalto-1                    Finland   
3                           AAUSat-4                    Denmark   
4                              ABS-2                         NR   

  Country of Operator/Owner                        Operator/Owner       Users  \
0                       USA                          Hera Systems  Commercial   
1                     Spain  Universitat Politècnica de Catalunya       Civil   
2                   Finland                      Aalto University       Civil   
3                   Denmark                 University of Aalborg       Civil   
4             Multinational         Asia Broadcast Satellite Ltd.  Commercial   

                  Purpose                       Detailed Purpose  \
0       Earth Observation                 Technology Development   
1  Technology Development                                    NaN   
2  Technology Development                                    NaN   
3       Earth Observation  Automatic Identification System (AIS)   
4          Communications                                    NaN   

  Class of Orbit       Type of Orbit  ...  Unnamed: 57  Unnamed: 58  \
0            LEO  Non-Polar Inclined  ...          NaN          NaN   
1            LEO                 NaN  ...          NaN          NaN   
2            LEO                 NaN  ...          NaN          NaN   
3            LEO     Sun-Synchronous  ...          NaN          NaN   
4            GEO                 NaN  ...          NaN          NaN   

   Unnamed: 59  Unnamed: 60  Unnamed: 61  Unnamed: 62  Unnamed: 63  \
0          NaN          NaN          NaN          NaN          NaN   
1          NaN          NaN          NaN          NaN          NaN   
2          NaN          NaN          NaN          NaN          NaN   
3          NaN          NaN          NaN          NaN          NaN   
4          NaN          NaN          NaN          NaN          NaN   

  Unnamed: 64 Unnamed: 65 Unnamed: 66  
0         NaN         NaN         NaN  
1         NaN         NaN         NaN  
2         NaN         NaN         NaN  
3         NaN         NaN         NaN  
4         NaN         NaN         NaN  

[5 rows x 67 columns]

## Limpieza de datos

In [200]:
satellites_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3375 entries, 0 to 3374
Data columns (total 67 columns):
 #   Column                              Non-Null Count  Dtype         
---  ------                              --------------  -----         
 0   Name of Satellite, Alternate Names  3372 non-null   object        
 1   Current Official Name of Satellite  3372 non-null   object        
 2   Country/Org of UN Registry          3372 non-null   object        
 3   Country of Operator/Owner           3372 non-null   object        
 4   Operator/Owner                      3372 non-null   object        
 5   Users                               3372 non-null   object        
 6   Purpose                             3372 non-null   object        
 7   Detailed Purpose                    916 non-null    object        
 8   Class of Orbit                      3372 non-null   object        
 9   Type of Orbit                       2751 non-null   object        
 10  Longitude of GEO (degree

Podemos observar que tenemos 3375 registros, de los cuales, en la mayoría de las columnas principales (0-6), tenemos 3372 valores no nulos. Exploremos esos tres valores faltantes

In [201]:
satellites_df.isna()

Name of Satellite, Alternate Names  Current Official Name of Satellite  \
0                                  False                               False   
1                                  False                               False   
2                                  False                               False   
3                                  False                               False   
4                                  False                               False   
...                                  ...                                 ...   
3370                               False                               False   
3371                               False                               False   
3372                                True                                True   
3373                                True                                True   
3374                                True                                True   

      Country/Org of UN Registry  Country of Operator/Owner  Operator/Owner  \
0                          False                      False           False   
1                          False                      False           False   
2                          False                      False           False   
3                          False                      False           False   
4                          False                      False           False   
...                          ...                        ...             ...   
3370                       False                      False           False   
3371                       False                      False           False   
3372                        True                       True            True   
3373                        True                       True            True   
3374                        True                       True            True   

      Users  Purpose  Detailed Purpose  Class of Orbit  Type of Orbit  ...  \
0     False    False             False           False          False  ...   
1     False    False              True           False           True  ...   
2     False    False              True           False           True  ...   
3     False    False             False           False          False  ...   
4     False    False              True           False           True  ...   
...     ...      ...               ...             ...            ...  ...   
3370  False    False             False           False          False  ...   
3371  False    False             False           False          False  ...   
3372   True     True              True            True           True  ...   
3373   True     True              True            True           True  ...   
3374   True     True              True            True           True  ...   

      Unnamed: 57  Unnamed: 58  Unnamed: 59  Unnamed: 60  Unnamed: 61  \
0            True         True         True         True         True   
1            True         True         True         True         True   
2            True         True         True         True         True   
3            True         True         True         True         True   
4            True         True         True         True         True   
...           ...          ...          ...          ...          ...   
3370         True         True         True         True         True   
3371         True         True         True         True         True   
3372         True         True         True         True         True   
3373         True         True         True         True         True   
3374         True         True         True         True         True   

      Unnamed: 62  Unnamed: 63  Unnamed: 64  Unnamed: 65  Unnamed: 66  
0            True         True         True         True         True  
1            True         True         True         True         True  
2            True         True         True         True         T

Podemos observar que los últimos tres registros son completamente nulos, por lo que los quitamos con la siguiente línea

In [202]:
satellites_df = satellites_df.iloc[:3372, :]

In [203]:
satellites_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3372 entries, 0 to 3371
Data columns (total 67 columns):
 #   Column                              Non-Null Count  Dtype         
---  ------                              --------------  -----         
 0   Name of Satellite, Alternate Names  3372 non-null   object        
 1   Current Official Name of Satellite  3372 non-null   object        
 2   Country/Org of UN Registry          3372 non-null   object        
 3   Country of Operator/Owner           3372 non-null   object        
 4   Operator/Owner                      3372 non-null   object        
 5   Users                               3372 non-null   object        
 6   Purpose                             3372 non-null   object        
 7   Detailed Purpose                    916 non-null    object        
 8   Class of Orbit                      3372 non-null   object        
 9   Type of Orbit                       2751 non-null   object        
 10  Longitude of GEO (degree

In [204]:
satellites_df["Source Used for Orbital Data"].to_frame().head()

Source Used for Orbital Data
0                JMSatcat/3_20
1                    Estimated
2               JMSatcat/10_17
3                      Space50
4                        ZARYA

In [205]:
l = [idx for idx, val in enumerate(satellites_df["Unnamed: 28"].notna()) if val]
satellites_df.loc[l, "Unnamed: 28"].to_frame().head()

Unnamed: 28
107  Imaging system that is to be tested in space f...
259     http://www.planet4589.org/space/log/satcat.txt
984

In [206]:
l = [idx for idx, val in enumerate(satellites_df["Unnamed: 60"].notna()) if val]
satellites_df.iloc[l, 60:]

Unnamed: 60  \
75                            http://www.orbcomm.com   
1323  http://perso.wanadoo.fr/eurospace/espdatabase/   
3080                          http://www.orbcomm.com   

                                           Unnamed: 61  \
75    http://nssdc.gsfc.nasa.gov/spacewarn/spx554.html   
1323  http://nssdc.gsfc.nasa.gov/spacewarn/spx540.html   
3080  http://nssdc.gsfc.nasa.gov/spacewarn/spx554.html   

                                            Unnamed: 62  Unnamed: 63  \
75    http://perso.wanadoo.fr/eurospace/espdatabase/...          NaN   
1323                                                NaN          NaN   
3080  http://perso.wanadoo.fr/eurospace/espdatabase/...          NaN   

      Unnamed: 64  Unnamed: 65  Unnamed: 66  
75            NaN          NaN  JMSatcat803  
1323          NaN          NaN          NaN  
3080          NaN          NaN  JMSatcat803

In [207]:
indices = [idx for idx, res in enumerate(satellites_df["Source.6"].notna()) if res]
satellites_df.loc[indices, "Source.6"].to_frame().head()

Source.6
10  http://www.astronautix.com/a/advancedorion.html
11  http://www.astronautix.com/a/advancedorion.html
12  http://www.astronautix.com/a/advancedorion.html
13  http://www.astronautix.com/a/advancedorion.html
14  http://www.astronautix.com/a/advancedorion.html

Después de analizar los valores faltantes, las columnas que tienen el mayor número de valores faltantes se refieres a columnas con fuentes sobre un satélite por lo que es seguro eliminar todas las columnas que involucren las fuente pues no será de relevancia para nuestro análisis. En este primer análisis nos quedamos con las columnas hasta el número NORAD.

In [208]:
satellites_df = satellites_df.iloc[:, :27]
print(satellites_df.shape)
satellites_df.head()

(3372, 27)


Name of Satellite, Alternate Names  \
0  1HOPSAT-TD (1st-generation High Optical Perfor...   
1                                             3Cat-1   
2                                           Aalto-1    
3                                           AAUSat-4   
4                           ABS-2 (Koreasat-8, ST-3)   

  Current Official Name of Satellite Country/Org of UN Registry  \
0                            1HOPSAT                 NR (12/20)   
1                             3Cat-1                         NR   
2                            Aalto-1                    Finland   
3                           AAUSat-4                    Denmark   
4                              ABS-2                         NR   

  Country of Operator/Owner                        Operator/Owner       Users  \
0                       USA                          Hera Systems  Commercial   
1                     Spain  Universitat Politècnica de Catalunya       Civil   
2                   Finland                      Aalto University       Civil   
3                   Denmark                 University of Aalborg       Civil   
4             Multinational         Asia Broadcast Satellite Ltd.  Commercial   

                  Purpose                       Detailed Purpose  \
0       Earth Observation                 Technology Development   
1  Technology Development                                    NaN   
2  Technology Development                                    NaN   
3       Earth Observation  Automatic Identification System (AIS)   
4          Communications                                    NaN   

  Class of Orbit       Type of Orbit  ...  Dry Mass (kg.)  Power (watts)  \
0            LEO  Non-Polar Inclined  ...             NaN            NaN   
1            LEO                 NaN  ...             NaN            NaN   
2            LEO                 NaN  ...             NaN            4.5   
3            LEO     Sun-Synchronous  ...             NaN            NaN   
4            GEO                 NaN  ...             NaN          16000   

   Date of Launch  Expected Lifetime (yrs.)  \
0      2019-12-11                       0.5   
1      2018-11-29                       NaN   
2      2017-06-23                       2.0   
3      2016-04-25                       NaN   
4      2014-02-06                      15.0   

                             Contractor  Country of Contractor  \
0                          Hera Systems                    USA   
1  Universitat Politècnica de Catalunya                  Spain   
2                      Aalto University                Finland   
3                 University of Aalborg                Denmark   
4                   Space Systems/Loral                    USA   

                  Launch Site Launch Vehicle COSPAR Number NORAD Number  
0  Satish Dhawan Space Centre           PSLV     2019-089H      44859.0  
1  Satish Dhawan Space Centre           PSLV     2018-096K      43728.0  
2  Satish Dhawan Space Centre           PSLV     2017-036L      42775.0  
3         Guiana Space Center     Soyuz 2.1a     2016-025E      41460.0  
4         Guiana Space Center   Ariane 5 ECA     2014-006A      39508.0  

[5 rows x 27 columns]

In [209]:
satellites_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3372 entries, 0 to 3371
Data columns (total 27 columns):
 #   Column                              Non-Null Count  Dtype         
---  ------                              --------------  -----         
 0   Name of Satellite, Alternate Names  3372 non-null   object        
 1   Current Official Name of Satellite  3372 non-null   object        
 2   Country/Org of UN Registry          3372 non-null   object        
 3   Country of Operator/Owner           3372 non-null   object        
 4   Operator/Owner                      3372 non-null   object        
 5   Users                               3372 non-null   object        
 6   Purpose                             3372 non-null   object        
 7   Detailed Purpose                    916 non-null    object        
 8   Class of Orbit                      3372 non-null   object        
 9   Type of Orbit                       2751 non-null   object        
 10  Longitude of GEO (degree

Adicionalmente, quitamos las siguiente columna que no aportan mucho a nuestro análisis
- *Name of Satellite, Alternate Names*: no nos interesa conocer el nombre más popular de un satélite ni nombres pasados, por lo que nos podemos quedar simplemente con la columna *Current Official Name of Satellite* como el identificador oficial de cada satélite 
- *Detailed Purpose*: columna con 916 valores no nulos de 3372 
- 

In [210]:
cols_to_drop = ["Name of Satellite, Alternate Names", "Detailed Purpose", "Longitude of GEO (degrees)", "Eccentricity", "Inclination (degrees)", "COSPAR Number", "NORAD Number", "Expected Lifetime (yrs.)", "Dry Mass (kg.)", "Power (watts)", "Launch Vehicle"]
satellites_df.drop(columns=cols_to_drop, inplace=True)
satellites_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3372 entries, 0 to 3371
Data columns (total 16 columns):
 #   Column                              Non-Null Count  Dtype         
---  ------                              --------------  -----         
 0   Current Official Name of Satellite  3372 non-null   object        
 1   Country/Org of UN Registry          3372 non-null   object        
 2   Country of Operator/Owner           3372 non-null   object        
 3   Operator/Owner                      3372 non-null   object        
 4   Users                               3372 non-null   object        
 5   Purpose                             3372 non-null   object        
 6   Class of Orbit                      3372 non-null   object        
 7   Type of Orbit                       2751 non-null   object        
 8   Perigee (km)                        3372 non-null   float64       
 9   Apogee (km)                         3372 non-null   float64       
 10  Period (minutes)        

In [211]:
satellites_df.rename(columns={
    "Current Official Name of Satellite": "Satellite",
    "Country/Org of UN Registry":"UN Country",
    "Country of Operator/Owner":"Country",
    "Class of Orbit":"Orbit Class",
    "Type of Orbit":"Orbit Type",
    "Perigee (km)":"Perigee",
    "Apogee (km)":"Apogee",
    "Period (minutes)":"Period",
    "Launch Mass (kg.)":"Launch Mass",
    "Country of Contractor":"Contractor Country",
    "Date of Launch":"Launch Date"
}, inplace=True)
satellites_df = satellites_df.reindex(columns=[
    "Satellite", "Country", "UN Country", "Operator/Owner", "Contractor", "Contractor Country", 
    "Users", "Purpose", "Launch Date", "Launch Site", "Orbit Class", "Orbit Type", 
    "Perigee", "Apogee", "Period", "Launch Mass"
])

In [212]:
print(satellites_df.shape)
satellites_df.head()

(3372, 16)


Satellite        Country  UN Country                        Operator/Owner  \
0   1HOPSAT            USA  NR (12/20)                          Hera Systems   
1    3Cat-1          Spain          NR  Universitat Politècnica de Catalunya   
2   Aalto-1        Finland     Finland                      Aalto University   
3  AAUSat-4        Denmark     Denmark                 University of Aalborg   
4     ABS-2  Multinational          NR         Asia Broadcast Satellite Ltd.   

                             Contractor Contractor Country       Users  \
0                          Hera Systems                USA  Commercial   
1  Universitat Politècnica de Catalunya              Spain       Civil   
2                      Aalto University            Finland       Civil   
3                 University of Aalborg            Denmark       Civil   
4                   Space Systems/Loral                USA  Commercial   

                  Purpose Launch Date                 Launch Site Orbit Class  \
0       Earth Observation  2019-12-11  Satish Dhawan Space Centre         LEO   
1  Technology Development  2018-11-29  Satish Dhawan Space Centre         LEO   
2  Technology Development  2017-06-23  Satish Dhawan Space Centre         LEO   
3       Earth Observation  2016-04-25         Guiana Space Center         LEO   
4          Communications  2014-02-06         Guiana Space Center         GEO   

           Orbit Type  Perigee   Apogee   Period  Launch Mass  
0  Non-Polar Inclined    566.0    576.0    96.08         22.0  
1                 NaN    476.0    500.0    95.00          4.0  
2                 NaN    497.0    517.0    94.70          4.5  
3     Sun-Synchronous    442.0    687.0    95.90          1.0  
4                 NaN  35778.0  35793.0  1436.03       6330.0

In [213]:
for c in sorted(satellites_df.Country.unique()): print(c)

Algeria
Argentina
Australia
Austria
Azerbaijan
Bangladesh
Belarus
Belgium
Bolivia
Brazil
Bulgaria
Canada
Chile
China
China/Brazil
China/France
China/Italy
Colombia
Czech Republic
Denmark
ESA
ESA/USA
ESA/USA/Russia
Ecuador
Egypt
Estonia
Ethiopia
Finland
France
France/Belgium/Sweden
France/Israel
France/Italy
France/Italy/Belgium/Spain/Greece
France/USA
Germany
Greece
Greece/United Kingdom
India
India/Canada
India/France
Indonesia
Iran
Iraq
Israel
Italy
Japan
Japan/Singapore
Jordan
Kazakhstan
Laos
Lithuania
Luxembourg
Malaysia
Mexico
Morocco
Morocco/Germany
Multinational
Nepal
Netherlands
Nigeria
Norway
Pakistan
Peru
Philippines
Poland
Qatar
Russia
Russia/USA
Saudi Arabia
Singapore
Singapore/Taiwan
Slovenia
South Africa
South Korea
Spain
Sri Lanka
Sudan
Sweden
Switzerland
Taiwan
Taiwan/USA
Thailand
Turkey
Turkmenistan/Monaco
USA
USA/Argentina
USA/Canada
USA/Canada/Japan
USA/France
USA/Germany
USA/Japan
USA/Japan/Brazil
USA/Mexico
USA/Sweden
USA/United Kingdom/Italy
Ukraine
United Arab Em